### PR TITLE
Generalize ``dd.to_datetime`` for GPU-backed collections, introduce ``get_meta_library`` utility

### LINKED ISSUE #9880: Generalize `dd.to_datetime` to support cuDF-backed Series
A couple months ago I opened Issue https://github.com/rapidsai/cudf/issues/12062 on cuDF, but after some discussions with @wence- @beckernick @charlesbluca it seems like it can be resolved on the Dask side.

**Minimal Complete Verifiable Example**:
```
import cudf
import dask_cudf
import dask.dataframe as dd

a = cudf.Series(['2022-01-01', '2022-02-02'])
dser = dask_cudf.from_cudf(a, 1)
res = dd.to_datetime(dser).compute()  # note that there are no errors if we don't call compute()
```
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In [19], line 6
      4 a = cudf.Series(['2022-01-01', '2022-02-02'])
      5 dser = dask_cudf.from_cudf(a, 1)
----> 6 res = dd.to_datetime(dser).compute()

File ~/miniconda3/envs/dsql_rapids-22.12/lib/python3.9/site-packages/dask/base.py:315, in DaskMethodsMixin.compute(self, **kwargs)
    291 def compute(self, **kwargs):
    292     """Compute this dask collection
    293 
    294     This turns a lazy Dask collection into its in-memory equivalent.
   (...)
    313     dask.base.compute
    314     """
--> 315     (result,) = compute(self, traverse=False, **kwargs)
    316     return result

File ~/miniconda3/envs/dsql_rapids-22.12/lib/python3.9/site-packages/dask/base.py:600, in compute(traverse, optimize_graph, scheduler, get, *args, **kwargs)
    597     keys.append(x.__dask_keys__())
    598     postcomputes.append(x.__dask_postcompute__())
--> 600 results = schedule(dsk, keys, **kwargs)
    601 return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])

File ~/miniconda3/envs/dsql_rapids-22.12/lib/python3.9/site-packages/dask/threaded.py:89, in get(dsk, keys, cache, num_workers, pool, **kwargs)
     86     elif isinstance(pool, multiprocessing.pool.Pool):
     87         pool = MultiprocessingPoolExecutor(pool)
---> 89 results = get_async(
     90     pool.submit,
     91     pool._max_workers,
     92     dsk,
     93     keys,
     94     cache=cache,
     95     get_id=_thread_get_id,
     96     pack_exception=pack_exception,
     97     **kwargs,
     98 )
    100 # Cleanup pools associated to dead threads
    101 with pools_lock:

File ~/miniconda3/envs/dsql_rapids-22.12/lib/python3.9/site-packages/dask/local.py:511, in get_async(submit, num_workers, dsk, result, cache, get_id, rerun_exceptions_locally, pack_exception, raise_exception, callbacks, dumps, loads, chunksize, **kwargs)
    509         _execute_task(task, data)  # Re-execute locally
    510     else:
--> 511         raise_exception(exc, tb)
    512 res, worker_id = loads(res_info)
    513 state["cache"][key] = res

File ~/miniconda3/envs/dsql_rapids-22.12/lib/python3.9/site-packages/dask/local.py:319, in reraise(exc, tb)
    317 if exc.__traceback__ is not tb:
    318     raise exc.with_traceback(tb)
--> 319 raise exc

File ~/miniconda3/envs/dsql_rapids-22.12/lib/python3.9/site-packages/dask/local.py:224, in execute_task(key, task_info, dumps, loads, get_id, pack_exception)
    222 try:
    223     task, data = loads(task_info)
--> 224     result = _execute_task(task, data)
    225     id = get_id()
    226     result = dumps((result, id))

File ~/miniconda3/envs/dsql_rapids-22.12/lib/python3.9/site-packages/dask/core.py:119, in _execute_task(arg, cache, dsk)
    115     func, args = arg[0], arg[1:]
    116     # Note: Don't assign the subtask results to a variable. numpy detects
    117     # temporaries by their reference count and can execute certain
    118     # operations in-place.
--> 119     return func(*(_execute_task(a, cache) for a in args))
    120 elif not ishashable(arg):
    121     return arg

File ~/miniconda3/envs/dsql_rapids-22.12/lib/python3.9/site-packages/dask/optimization.py:990, in SubgraphCallable.__call__(self, *args)
    988 if not len(args) == len(self.inkeys):
    989     raise ValueError("Expected %d args, got %d" % (len(self.inkeys), len(args)))
--> 990 return core.get(self.dsk, self.outkey, dict(zip(self.inkeys, args)))

File ~/miniconda3/envs/dsql_rapids-22.12/lib/python3.9/site-packages/dask/core.py:149, in get(dsk, out, cache)
    147 for key in toposort(dsk):
    148     task = dsk[key]
--> 149     result = _execute_task(task, cache)
    150     cache[key] = result
    151 result = _execute_task(out, cache)

File ~/miniconda3/envs/dsql_rapids-22.12/lib/python3.9/site-packages/dask/core.py:119, in _execute_task(arg, cache, dsk)
    115     func, args = arg[0], arg[1:]
    116     # Note: Don't assign the subtask results to a variable. numpy detects
    117     # temporaries by their reference count and can execute certain
    118     # operations in-place.
--> 119     return func(*(_execute_task(a, cache) for a in args))
    120 elif not ishashable(arg):
    121     return arg

File ~/miniconda3/envs/dsql_rapids-22.12/lib/python3.9/site-packages/dask/utils.py:71, in apply(func, args, kwargs)
     40 """Apply a function given its positional and keyword arguments.
     41 
     42 Equivalent to ``func(*args, **kwargs)``
   (...)
     68 >>> dsk = {'task-name': task}  # adds the task to a low level Dask task graph
     69 """
     70 if kwargs:
---> 71     return func(*args, **kwargs)
     72 else:
     73     return func(*args)

File ~/miniconda3/envs/dsql_rapids-22.12/lib/python3.9/site-packages/dask/dataframe/core.py:6770, in apply_and_enforce(*args, **kwargs)
   6768 func = kwargs.pop("_func")
   6769 meta = kwargs.pop("_meta")
-> 6770 df = func(*args, **kwargs)
   6771 if is_dataframe_like(df) or is_series_like(df) or is_index_like(df):
   6772     if not len(df):

File ~/miniconda3/envs/dsql_rapids-22.12/lib/python3.9/site-packages/pandas/core/tools/datetimes.py:1100, in to_datetime(arg, errors, dayfirst, yearfirst, utc, format, exact, unit, infer_datetime_format, origin, cache)
   1098         result = _convert_and_box_cache(argc, cache_array)
   1099     else:
-> 1100         result = convert_listlike(argc, format)
   1101 else:
   1102     result = convert_listlike(np.array([arg]), format)[0]

File ~/miniconda3/envs/dsql_rapids-22.12/lib/python3.9/site-packages/pandas/core/tools/datetimes.py:413, in _convert_listlike_datetimes(arg, format, name, tz, unit, errors, infer_datetime_format, dayfirst, yearfirst, exact)
    410         return idx
    411     raise
--> 413 arg = ensure_object(arg)
    414 require_iso8601 = False
    416 if infer_datetime_format and format is None:

File pandas/_libs/algos_common_helper.pxi:33, in pandas._libs.algos.ensure_object()

File ~/miniconda3/envs/dsql_rapids-22.12/lib/python3.9/site-packages/cudf/core/frame.py:447, in Frame.__array__(self, dtype)
    446 def __array__(self, dtype=None):
--> 447     raise TypeError(
    448         "Implicit conversion to a host NumPy array via __array__ is not "
    449         "allowed, To explicitly construct a GPU matrix, consider using "
    450         ".to_cupy()\nTo explicitly construct a host matrix, consider "
    451         "using .to_numpy()."
    452     )

TypeError: Implicit conversion to a host NumPy array via __array__ is not allowed, To explicitly construct a GPU matrix, consider using .to_cupy()
To explicitly construct a host matrix, consider using .to_numpy().
```

**Anything else we need to know?**:
@beckernick traced `to_datetime()` [here](https://github.com/dask/dask/blob/23a75632b7644f1289a760031d9ba85a35cbfd20/dask/dataframe/core.py#L7845) and suggested to generalize the Dask implementation by potentially using [this pattern](https://github.com/dask/dask/blob/23a75632b7644f1289a760031d9ba85a35cbfd20/dask/dataframe/reshape.py#L154-L155).

### PR DESCRIPTION
**Pull request by charlesbluca**
This PR should unblock some datetime related work going on in https://github.com/dask-contrib/dask-sql/pull/984, but also hopefully can be a place to host a wider discussion around what a utility API to get the top-level module of a Dask collection's partition type should look like.

cc @rjzamora @sarahyurick

- [ ] Closes #9880
- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

**Review by rjzamora — COMMENTED** [review]
I think this utility makes sense @charlesbluca - Thanks!

My only preference would be to replace the term `serial` with `meta` throughout.  I think this would make it clearer that we are extracting the module name for collection's current "meta" property.  This naming would also be more consistent with the existing `meta_lib_from_array` dispatch.

**Side note**: I agree with your decision to use a utility instead of a dispatch function (like `meta_lib_from_array`). The same logic already works for cudf, pandas, cupy and numpy. Therefore, there is really no reason to over-engineer a dispatch-based solution for this.

**Review by jrbourbeau — COMMENTED** [review]
+1 this seems reasonable to me. I can think of other places where this utility would be useful. Also agree with @rjzamora about the name change to something like `get_meta_library`

File: dask/utils.py [hunk 1]
```diff
@@ -2094,3 +2094,10 @@ def get_default_shuffle_algorithm() -> str:
         return "tasks"
     except (ImportError, ValueError):
         return "disk"
+
+
+def get_meta_library(like):
```
Comment (jrbourbeau) [inline dask/utils.py:2099]: Could you add some tests for this?

File: dask/dataframe/core.py [hunk 2]
```diff
@@ -7819,9 +7820,12 @@ def maybe_shift_divisions(df, periods, freq):
 @wraps(pd.to_datetime)
 def to_datetime(arg, meta=None, **kwargs):
     tz_kwarg = {"tz": "utc"} if kwargs.get("utc") else {}
+
+    xd = get_meta_library(arg)
```
Comment (jrbourbeau) [inline dask/dataframe/core.py:7824]: `dask/dataframe/tests/test_dataframe.py::test_to_datetime` is failing here. It looks like we might need to move this line of code around

**Review by jrbourbeau — COMMENTED** [review]
Thanks @charlesbluca!

File: continuous_integration/environment-3.9.yaml [hunk 3]
```diff
@@ -1,6 +1,7 @@
 # This job includes coverage
 name: test-environment
 channels:
+  - rapidsai-nightly
```
Comment (jrbourbeau) [inline continuous_integration/environment-3.9.yaml:4]: My guess is this isn't needed (but do let me know if I'm missing something) 

```suggestion
```

**Review by jrbourbeau — COMMENTED** [review]
Thanks @charlesbluca! Apologies for the delayed response. 

Overall this looks good. I have one minor question about adding the `rapidsai-nightly` channel, but otherwise this PR could be merged as is. 

A couple of non-blocking comments:

1. Have you confirmed the changes here close https://github.com/dask/dask/issues/9880? It might be good to add a test
2. `test_get_meta_library` covers the usual, `numpy` / `pandas` backed cases. It might be worth adding a second test that `cupy` / `cudf` backed collections.

File: dask/dataframe/core.py [hunk 4]
```diff
@@ -8013,9 +8014,11 @@ def maybe_shift_divisions(df, periods, freq):
 @wraps(pd.to_datetime)
 def to_datetime(arg, meta=None, **kwargs):
     tz_kwarg = {"tz": "utc"} if kwargs.get("utc") else {}
+
     if meta is None:
         if isinstance(arg, Index):
-            meta = pd.DatetimeIndex([], **tz_kwarg)
+            xd = get_meta_library(arg)
+            meta = xd.DatetimeIndex([], **tz_kwarg)
```
Comment (rjzamora) [inline dask/dataframe/core.py:8021]: Just a nit, but is there any good reason for creating the `xd` variable here (and below)?

File: dask/dataframe/reshape.py [hunk 5]
```diff
@@ -151,11 +149,10 @@ def get_dummies(
         if not all(has_known_categories(data[c]) for c in columns):
             raise NotImplementedError(unknown_cat_msg)
 
-    package_name = data._meta.__class__.__module__.split(".")[0]
-    dummies = sys.modules[package_name].get_dummies
+    xd = get_meta_library(data)
 
     return map_partitions(
-        dummies,
+        xd.get_dummies,
```
Comment (rjzamora) [inline dask/dataframe/reshape.py:155]: Similar comment: Seems like it should be fine to just use `get_meta_library(data).get_dummies` in the call to `map_partitions`?

**Review by rjzamora — COMMENTED** [review]
Agree with @jrbourbeau that we should test that this resolves the motivating issue. Otherwise, this looks good. Only had a nit-pick suggestion.

File: continuous_integration/environment-3.9.yaml (hunk 3 above)
Comment (charlesbluca) [reply continuous_integration/environment-3.9.yaml:4]: Yup, thanks! Missed this while doing some local testing

File: dask/dataframe/core.py (hunk 4 above)
Comment (charlesbluca) [reply dask/dataframe/core.py:8021]: None that I can think of; will go with your below suggestion for both cases

**Comment by charlesbluca** [comment]
Looks like the gpuCI failures here are due to differences in datetime handling between cuDF and pandas - going to see what can be done upstream before making any more updates

File: dask/dataframe/core.py [hunk 6]
```diff
@@ -8034,31 +8035,51 @@ def maybe_shift_divisions(df, periods, freq):
...
+                    )
+                else:
+                    arg_meta = arg._meta_nonempty
+                    assign_dict = dict()
+                    for col in arg_meta.columns:
+                        if arg_meta[col].dtype.kind == "O":
+                            assign_dict[col] = "2000" if col == "year" else "1"
+                        elif col == "year" and is_numeric_dtype(arg_meta[col]):
+                            assign_dict[col] = 2000
+                    arg_meta = arg_meta.assign(**assign_dict)
```
Comment (jrbourbeau) [inline dask/dataframe/core.py:8077]: This looks complicated -- why is this change needed?

File: dask/dataframe/core.py [hunk 7]
```diff
@@ -8034,31 +8035,51 @@ def maybe_shift_divisions(df, periods, freq):
...
+            meta = get_meta_library(arg).to_datetime(arg_meta, **kwargs)
```
Comment (jrbourbeau) [inline dask/dataframe/core.py:8078]: Previously we were passing `tz_kwarg` here. Is that no longer needed?

**Review by jrbourbeau — COMMENTED** [review]
Thanks @charlesbluca. Overall this looks good. Just had a couple of final questions about the `to_datetime`-specific changes

File: dask/dataframe/core.py (hunk 7 above)
Comment (charlesbluca) [reply dask/dataframe/core.py:8078]: From what I understand, `tz_kwarg` was used here to grab only the relevant `to_datetime` kwarg for the Series constructor - now that we're constructing the meta directly with `to_datetime`, we should be good to pass all of these kwargs directly in (with the exception of `infer_datetime_format` for pandas 2)

File: dask/dataframe/core.py (hunk 6 above)
Comment (charlesbluca) [reply dask/dataframe/core.py:8077]: `cudf.to_datetime` has a somewhat nuanced handling of datetime precision compared to pandas that made it so the only way to get the correct meta here was by directly calling `xd.to_datetime` on some subset of the data. This gets a little hairy when using `_meta_nonempty`, which generates data that would error in `to_datetime` (string columns containing non-integers, int columns containing out of range years, etc.), which this code attempts to resolve.

As I'm writing this, I wonder if maybe a simpler solution here would be to disable dtype checking on the Dask object in the tests? Even with this code, there are still some cases I can think of where the dtypes of the meta and the computed frame wouldn't match up (at this point, we don't know if any of the columns contain timezone-aware dates), and I'd imagine we'd prefer keeping the original meta construction which, though inaccurate in a few cases, isn't likely to introduce a breaking change.

Do you have a preference here?

Comment (charlesbluca) [reply dask/dataframe/core.py]: Ended up going back to the original meta construction and loosening dtype checks on GPU tests - do we think it's worth emitting a warning here if we run into that case?

**Comment by j-bennet** [comment]
@charlesbluca This has been sitting for a while, but it looks like all the reviews were 👍 . Is there any reason not to merge it?

**Review by jrbourbeau — APPROVED** [review]
Thanks for the ping @j-bennet. I agree with @charlesbluca, this looks good to go 👍
